zip-0155.rst

```diff
@@ -0,0 +1,246 @@
+  +--------------+-----------------+---------------------+----------------------------------------------------------------+
+  | ``varies``   | ``sizeAddr``    | ``CompactSize``     | The length in bytes of ``addr``.                               |
+  +--------------+-----------------+---------------------+----------------------------------------------------------------+
+  | ``sizeAddr`` | ``addr``        | ``uint8[sizeAddr]`` | Network address. The interpretation depends on networkID.      |
+  +--------------+-----------------+---------------------+----------------------------------------------------------------+
+  | 2            | ``port``        | ``uint16``          | Network port. If not relevant for the network this MUST be 0.  |
+  +--------------+-----------------+---------------------+----------------------------------------------------------------+
+
+One message can contain up to 1,000 addresses. Clients MUST reject messages with
+more addresses.
+
+Field ``addr`` has a variable length, with a maximum of 512 bytes (4096 bits). Clients
+SHOULD reject messages with longer addresses, irrespective of the network ID.
+
+The list of reserved network IDs is as follows:
+
+  +------------+-------------+------------------------+-----------------------------------------+
+  | Network ID | Enumeration | Address length (bytes) | Description                             |
+  +------------+-------------+------------------------+-----------------------------------------+
+  | ``0x01``   | ``IPV4``    | 4                      | IPv4 address (globally routed internet) |
+  +------------+-------------+------------------------+-----------------------------------------+
+  | ``0x02``   | ``IPV6``    | 16                     | IPv6 address (globally routed internet) |
+  +------------+-------------+------------------------+-----------------------------------------+
+  | ``0x04``   | ``TORV3``   | 32                     | Tor v3 hidden service address           |
+  +------------+-------------+------------------------+-----------------------------------------+
+  | ``0x05``   | ``I2P``     | 32                     | I2P overlay network address             |
+  +------------+-------------+------------------------+-----------------------------------------+
+  | ``0x06``   | ``CJDNS``   | 16                     | Cjdns overlay network address           |
+  +------------+-------------+------------------------+-----------------------------------------+
+
+Network ID ``0x03`` is intentionally not assigned. In BIP 155 [#bip-0155]_ it was
+assigned to Tor v2 hidden service addresses, but those addresses are no longer
+supported by the latest Tor client code, and MUST NOT be used once this ZIP is
+deployed.
+
+Clients SHOULD gossip addresses from all known networks even if they are currently
+not connected to some of them. That could help multi-homed nodes and make it more
+difficult for an observer to tell which networks a node is connected to.
```

Comment:
What should clients do about known but un-validated addresses?
(For example, a client that is based on this spec, but doesn't check that tor onion service addresses are valid ed25519 keys.)

The above paragraph suggests that they should gossip them, but the below paragraph says that we don't want to gossip un-validated addresses.

Reply:
Note to self: check https://github.com/bitcoin/bitcoin/blob/master/src/netaddress.cpp

Reply:
This comment is suggestive:

https://github.com/bitcoin/bitcoin/blob/92f3a4b4d0531976c9484f5b5075487e0925b619/src/netaddress.cpp#L149

       // TORv2-in-IPv6 (unsupported). Unserialize as !IsValid(), thus ignoring them.
       // Mimic a default-constructed CNetAddr object which is !IsValid() and thus
       // will not be gossiped, but continue reading next addresses from the stream.

I'm examining `IsValid`.

Reply:
https://github.com/bitcoin/bitcoin/blob/4fc15d15667d9d9c4fb5515ce73c05b4596298ec/src/netaddress.h#L223

Reply:
I changed this to "Clients SHOULD gossip valid, potentially routable addresses...". It is up to a node implementation which addresses on each network it considers valid and potentially routable.